docs: add EDK details and context around switching to utilities (#7239)
meltano/meltano: docs/src/_guide/advanced-topics.md

* `gcs` - Support for using Google Cloud Storage as a [state backend](/concepts/state_backends)
* `azure` - Support for using Azure Blob Storage as a [state backend](/concepts/state_backends)

## Extension Developer Kit (EDK)

> Meltano extensions are lightweight shims that allow you to integrate existing data tools with Meltano.

Reviewer: Shim can often mean something different. A different word here could be more accurate. Something like...

```suggestion
Meltano extensions are lightweight executables which allow you to integrate existing data tools with Meltano.
```

```suggestion
Meltano extensions are lightweight programs which allow you to integrate existing data tools with Meltano.
```

> Meltano extensions are thin wrappers around applications that allow you to integrate existing data tools with Meltano.

Reviewer 2: ```suggestion
Meltano extensions are lightweight shims that allow you to integrate existing data tools with Meltano.
```

Thoughts on this language? I don't want to give the impression that we fully wrap the tool and that our Airflow is somehow different from a "pure" Airflow

> Extensions allow the developer to add additional features like pre and post hooks to run before and after Meltano executes the wrapped application.

Reviewer 2: ```suggestion
Extensions allow the developer to add additional features like pre/post-hooks to run before/after Meltano executes the application, as well enabling project scaffolding to be customized for each plugin. This project scaffolding was previously accomplished via files bundles.
```

The Meltano [Extension Developer Kit](https://github.com/meltano/edk) (EDK) was created to make it easier for developers to build Meltano extensions.
For more information on how to build your own EDK-based Meltano extension, see [the EDK docs](https://edk.meltano.com/en/latest/).

> ### EDK Context

Reviewer: This section is more about how Meltano is classifying plugins and less about the EDK itself - as the EDK itself is agnostic to the plugin type we place upon it.

```suggestion
### Meltano Plugin Types
```

Or something like

```suggestion
### Meltano Utilities
```

Or

```suggestion
### Selecting a Plugin Type
```

Or

```suggestion
### Understanding Plugin Types
```

> The Meltano [Extension Developer Kit](https://github.com/meltano/edk) (EDK) was created to make it easier for developers to build Meltano extensions.
> The EDK also has functionality to allow the extension to initialize the scaffold of a project directory which was traditionally the job of files bundles, ultimately simplifying the work needed for developers.
> For more information on how to build your own EDK-based Meltano extension, see [the EDK docs](https://edk.meltano.com/en/latest/).

Reviewer 2: ```suggestion

The Meltano [Extension Developer Kit](https://github.com/meltano/edk) (EDK) was created to make it easier for developers to build Meltano extensions.
For more information on how to build your own EDK-based Meltano extension, see [the EDK docs](https://edk.meltano.com/en/latest/).
```

Meltano has traditionally assigned a plugin type to each plugin based on their functionality.
These plugin types were/are used in the Meltano codebase to activate plugin type specific features (i.e. piping Singer taps and targets together, running dbt deps before each run, or compiling and removing the Airflow config.cfg configuration file to avoid storing sensitive credentials).
This approach caused some challenges around getting new features implemented and accepted by the entire user base of Meltano because only one implementation was allowed:

* extractor (e.g. tap-github)
* loader (e.g. target-snowflake)
* transformer (e.g. dbt-snowflake)
* orchestrator (e.g. Airflow, Dagster)
* utility (e.g. SqlFluff)

The new approach is to group all non-EL plugins under the utility plugin type, leaving only the following plugin types:

* extractor
* loader

> * utility - now also including transformers and orchestrators

Reviewer 2: ```suggestion
* utility - now also including plugins previously called transformers and orchestrators
```
